### PR TITLE
disk spilling support for hash probe

### PR DESCRIPTION
**Pull request by xiaoxmeng**
The added hash probe procedure works as follow:
1. hash probe operator waits for table built by hash build operators
    to join
2. hash probe operator prepare for spilling: (1) setup spiller to spill
    input rows which corresponding partitions have been spilled at the
    build side.
4. hash probe operator process inputs and joins with the table
5. after all the hash probe operators have finished processing inputs,
    the last operator will do last probe if the join type needs, and
    notify the hash build operators of the next table to build from the
    spill data if there is remaining spill data to restore.
    Otherwise finish the probe operator processing.
7. Go back to step1 to start the next round of hash probe processing.
    
Unit tests are added in HashJoinTest to make each existing test case
to run test with spilling and different spilling levels.

This PR also fixes the issue that clears the output row map when the
build side table is empty detected by left join types when running
recursive spilling.

The followup will add spilling stats for HashProbe

### CODE REVIEW

File: velox/exec/HashTable.h [hunk 1]
```diff
@@ -89,6 +89,11 @@ class BaseHashTable {
   struct RowsIterator {
     int32_t hashTableIndex_{-1};
     RowContainerIterator rowContainerIterator_;
+
+    void reset() {
```
Comment (Yuhta) [inline velox/exec/HashTable.h:93]: Just `lastProbeIterator_ = {}` should enough?

File: velox/exec/HashProbe.h [hunk 2]
```diff
@@ -30,13 +31,45 @@ class HashProbe : public Operator {
...
+  ///
+  ///                           +--------------------------------+
+  ///                           ^                                |
+  ///                           |                                V
+  ///   kWaitForBuild -->  kRunning  -->  kWaitForPeers --> kFinish
+  ///                          ^               |
+  ///                          |               v
+  ///                          +---------------+
+  ///
+  enum class State {
```
Comment (Yuhta) [inline velox/exec/HashProbe.h:45]: nit: Probably should be `private`?

**Review by Yuhta — COMMENTED** [review]
Just some quick points, I will leave it to Masha or Orri for a more thorough review.

File: velox/exec/HashTable.h (hunk 1 above)
Comment (xiaoxmeng) [reply velox/exec/HashTable.h:93]: I think reset is better then assignment.

File: velox/exec/HashProbe.h (hunk 2 above)
Comment (xiaoxmeng) [reply velox/exec/HashProbe.h:45]: Yeah, but need to overload << operator for this class. Can we overload it for a private class?

Comment (Yuhta) [reply velox/exec/HashProbe.h:45]: It's probably fine to leave it as public then

File: velox/exec/HashTable.h (hunk 1 above)
Comment (Yuhta) [reply velox/exec/HashTable.h:93]: You would need to repeat the default value in two places.  Does not hurt too much for a small class like this though.
There is a way to get best of both worlds: `void reset() { *this = {}; }`.

Comment (xiaoxmeng) [reply velox/exec/HashTable.h:93]: I mean reset() is more clear in such use case than an assignment to default. Yeah, we could implement reset as *this = {}. Will update. Thanks!

**Comment by mbasmanova** [comment]
@xiaoxmeng This PR is very large, > 1K lines. Would it be possible to break it up into smaller pieces? Can the 3 fixed for pre-existing issues listed in "This PR also fixes the following issues:" be extracted into separate PRs? Smaller PRs are easier and quicker to review.

**Comment by mbasmanova** [comment]
> 5. Go back to step1 to start the next round of hash probe processing.

Thank you for writing a description of the workflow. It is nice and helps reviewing. It will be also helpful for future readers who can access these descriptions from git log and git blame (as commit messages). 

Step 5 is not exactly accurate, because step 2 applies only to the initial processing of the probe rows. The first pass is processing inputs and spill some of these while joining others. Subsequent passes read previously spilled rows and process all of these without spilling some. It would be nice to clarify.

File: velox/common/file/File.cpp [hunk 3]
```diff
@@ -75,20 +75,28 @@ LocalReadFile::LocalReadFile(std::string_view path) {
...
       rc,
       path,
-      strerror(errno));
+      folly::errnoStr(errno));
   size_ = rc;
 }
 
 LocalReadFile::LocalReadFile(int32_t fd) : fd_(fd) {}
 
+LocalReadFile::~LocalReadFile() {
```
Comment (mbasmanova) [inline velox/common/file/File.cpp:92]: It would be nice to extract this fix into a separate PR.

File: velox/common/file/tests/FileTest.cpp [hunk 4]
```diff
@@ -149,3 +155,35 @@ TEST(LocalFile, list) {
   localFs->remove(b);
   ASSERT_TRUE(localFs->list(std::string_view(tempFolder->path)).empty());
 }
+
+TEST(LocalFile, readFileDestructor) {
```
Comment (mbasmanova) [inline velox/common/file/tests/FileTest.cpp:159]: Thank you for adding a test.

File: velox/core/Context.h [hunk 5]
```diff
@@ -265,7 +265,6 @@ class BaseConfigManager {
     return SubscriptionHandle{typeid(T), subscription};
   }
 
-  //  std::atomic<shared_ptr<const Config>> config_;
```
Comment (mbasmanova) [inline velox/core/Context.h:268]: Thank you for the cleanup.

File: velox/core/QueryConfig.h [hunk 6]
```diff
@@ -282,6 +291,10 @@ class QueryConfig {
     return get<int32_t>(kTestingSpillPct, 0);
   }
 
+  int32_t maxSpillLevel() const {
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:294]: Nice. Thank you for adding this configuration property. It would be nice to extract this change and corresponding logic in HashBuild into a separate PR.

File: velox/exec/Driver.h [hunk 7]
```diff
@@ -126,8 +126,12 @@ enum class BlockingReason {
   kWaitForSplit,
   kWaitForExchange,
   kWaitForJoinBuild,
-  /// Build operator is blocked waiting for the probe operators to finish
-  /// probing before build the next hash table from the previously spilled data.
+  /// For a build operator, it is blocked waiting for the probe operators to
```
Comment (mbasmanova) [inline velox/exec/Driver.h:129]: Nice comment. It would be great to add similar comments to all enum values (in a separate PR).

File: velox/exec/HashBuild.cpp [hunk 8]
```diff
@@ -166,15 +169,26 @@ void HashBuild::setupSpiller(SpillPartition* spillPartition) {
...
+  if (spillPartition == nullptr) {
+    spillGroup_->addOperator(
+        *this,
+        [&](const std::vector<Operator*>& operators) { runSpill(operators); });
+  } else {
+    spillInputReader_ = spillPartition->createReader();
+
     const auto startBit = spillPartition->id().partitionBitOffset() +
         spillConfig.hashBitRange.numBits();
+    // Disable spilling if exceeding the max spill level.
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:184]: Curious what is the error message on exceeding spill level. Does it explain what happened clearly?

File: velox/exec/HashBuild.cpp [hunk 9]
```diff
@@ -700,9 +709,9 @@ bool HashBuild::finishHashBuild() {
     if (spiller_ != nullptr) {
       spillStats += spiller_->stats();
 
-      stats_.spilledBytes = spillStats.spilledBytes;
-      stats_.spilledRows = spillStats.spilledRows;
-      stats_.spilledPartitions = spillStats.spilledPartitions;
+      stats_.spilledBytes += spillStats.spilledBytes;
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:712]: Is this a fix for a pre-existing issue?

File: velox/exec/HashBuild.cpp [hunk 10]
```diff
@@ -812,6 +821,13 @@ void HashBuild::addRuntimeStats() {
           fmt::format("distinctKey{}", i), RuntimeCounter(asDistinct));
     }
   }
+  // Add max spilling level stats if spilling has been triggered.
+  if (spiller_ != nullptr && spiller_->isAnySpilled()) {
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:825]: It would be nice to document this new runtime stat in https://facebookincubator.github.io/velox/develop/joins.html#execution-statistics

File: velox/exec/HashTable.h [hunk 11]
```diff
@@ -89,6 +89,10 @@ class BaseHashTable {
...
+      *this = {};
```
Comment (mbasmanova) [inline velox/exec/HashTable.h:94]: This is an unusual syntax which may confuse future readers. Any particular reason not to explicitly reassign member variables? Perhaps, add a comment to explain what this does?

File: velox/exec/Spiller.cpp [hunk 12]
```diff
@@ -569,6 +569,24 @@ std::string Spiller::toString() const {
       spillFinalized_);
 }
 
+int32_t Spiller::Config::spillLevel(uint8_t startBitOffset) const {
+  VELOX_CHECK_LE(startBitOffset + hashBitRange.numBits(), 64);
```
Comment (mbasmanova) [inline velox/exec/Spiller.cpp:573]: nit: perhaps, include startBitOffset and hashBitRange.numBits() in the error message for easier troubleshooting

File: velox/exec/Spiller.cpp [hunk 13]
```diff
@@ -569,6 +569,24 @@ std::string Spiller::toString() const {
...
+  const int32_t deltaBits = startBitOffset - hashBitRange.begin();
+  VELOX_CHECK_GE(deltaBits, 0);
+  VELOX_CHECK_EQ(deltaBits % hashBitRange.numBits(), 0);
```
Comment (mbasmanova) [inline velox/exec/Spiller.cpp:576]: nit: perhaps, include deltaBits and hashBitRange.numBits() in the error message

nit: consider introducing a temp variable to avoid repeating hashBitRange.numBits()

File: velox/core/QueryConfig.h [hunk 14]
```diff
@@ -135,6 +135,15 @@ class QueryConfig {
 
   static constexpr const char* kTestingSpillPct = "testing.spill-pct";
 
+  /// The max allowed spilling level with zero being the initial spilling level.
+  /// This only applies for hash build spilling which might trigger recursive
+  /// spilling when the build table is too big. If it is set to -1, then there
```
Comment (mbasmanova) [inline velox/core/QueryConfig.h:140]: Should we provide some guideline on how to set this value? Would the user understand the implicit limit and choose this value to be less than that?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng I haven't finished reading. Still need to read HashProbe.h/cpp. Flushing some initial comments.

File: velox/exec/HashProbe.h [hunk 15]
```diff
@@ -30,13 +31,44 @@ class HashProbe : public Operator {
...
   bool needsInput() const override {
-    if (finished_ || noMoreInput_ || input_) {
+    if (state_ == State::kFinish || noMoreInput_ || noMoreSpillInput_ ||
+        input_ != nullptr) {
       return false;
     }
     if (table_) {
       return true;
     }
+    // NOTE: if we can't apply dynamic filtering, then we can start early to
```
Comment (mbasmanova) [inline velox/exec/HashProbe.h:70]: That's a good point. Perhaps, turn this into a TODO.

File: velox/exec/HashProbe.h [hunk 16]
```diff
@@ -55,7 +87,26 @@ class HashProbe : public Operator {
   void clearDynamicFilters() override;
 
  private:
-  // Sets up 'filter_' and related members.
+  void setState(State state);
+  void stateTransitionCheck(State state);
```
Comment (mbasmanova) [inline velox/exec/HashProbe.h:91]: nit: rename to checkStateTransition and make it const to be consistent with checkRunning() method below

File: velox/exec/HashProbe.h [hunk 17]
```diff
@@ -55,7 +87,26 @@ class HashProbe : public Operator {
...
+
+  void setRunning();
+  void checkRunning() const;
+  bool isRunning() const;
+
+  void waitForTableBuild();
```
Comment (mbasmanova) [inline velox/exec/HashProbe.h:97]: nit: waitForHashTable?

File: velox/exec/HashProbe.h [hunk 18]
```diff
@@ -55,7 +87,26 @@ class HashProbe : public Operator {
...
+
+  // Invoked to set up spilling related input processing. The function sets up a
+  // reader to read probe inputs from spilled data on disk if
+  // 'restoredSpillPartitionId' is not null. If 'spillPartitionIds' is not
+  // empty, then spilling has been triggered at the build side and the function
+  // will set up a spiller and the associated data structures to spill probe
+  // inputs.
+  void maybeSetupSpillInput(
```
Comment (mbasmanova) [inline velox/exec/HashProbe.h:105]: I assume that only one of the parameters can be non-empty, e.g. it is not possible both restoredSpillPartitionId and spillPartitionIds to be non-empty. Is this so? Would be nice to clarify in the comments and perhaps check in the implementation.

File: velox/exec/HashProbe.cpp [hunk 19]
```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
...
-BlockingReason HashProbe::isBlocked(ContinueFuture* future) {
-  if (table_) {
-    return BlockingReason::kNotBlocked;
+void HashProbe::maybeSetupSpillInput(
+    const std::optional<SpillPartitionId>& restoredPartitionId,
+    const SpillPartitionIdSet& spillPartitionIds) {
+  VELOX_CHECK_NULL(spillInputReader_);
+
+  // If 'restoredPartitionId' is not null, then 'table_' is built from the
+  // spilled build data. We need to create an unsorted reader to read the probe
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:234]: nit: perhaps, "We need to create" -> "Create ..."

File: velox/exec/HashProbe.cpp [hunk 20]
```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
...
+    spillInputReader_ = partition->createReader();
+    spillPartitionSet_.erase(iter);
   }
 
-  auto hashBuildResult =
-      operatorCtx_->task()
-          ->getHashJoinBridge(
-              operatorCtx_->driverCtx()->splitGroupId, planNodeId())
-          ->tableOrFuture(future);
+  VELOX_CHECK_NULL(spiller_);
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:245]: nit: would be nice to add a comment for this block of code to clarify that previous blocks sets up reading previously spilled partition while this block sets up spilling for the incoming data.

File: velox/exec/HashProbe.cpp [hunk 21]
```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
...
+  spillInputPartitionIds_ = spillPartitionIds;
+  if (!spillInputPartitionIds_.empty()) {
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:247]: nit: perhaps, return early if spillInputPartitionIds_ is empty to reduce the indentation size of the code block.

File: velox/exec/HashProbe.h [hunk 22]
```diff
@@ -108,11 +159,83 @@ class HashProbe : public Operator {
...
+  // remaining to restore. The function will reset the internal states which
+  // are relevant to the last finished probe run. The last finished probe
+  // operator will also notify the hash build operators to build the next hash
+  // table from spilled data.
+  void prepareForSpillRestore();
+
+  // Invoked to read next batch of spilled probe inputs from disk to process.
+  void addSpillInput();
+
+  void spillInput(RowVectorPtr& input);
```
Comment (mbasmanova) [inline velox/exec/HashProbe.h:199]: Would be nice to add comments for this method and two methods after it.

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Finished reading HashProbe.h. Thank you for adding detailed comments. Some questions / comments. Will read HashProbe.cpp next.

File: velox/exec/HashTable.h (hunk 11 above)
Comment (mbasmanova) [reply velox/exec/HashTable.h:94]: Makes sense. Thank you for clarifying. I don't think we use this syntax elsewhere in the codebase, hence, I thought it might be surprising to readers. Should we start using this more? For example, in OperatorStats::clear.

Comment (Yuhta) [reply velox/exec/HashTable.h:94]: I think we should use it wherever possible (i.e. struct with default constructor that takes no argument).  For `OperatorStats`, there is no default constructor, so the cleaner solution would be calling `clear()` inside constructor and remove the default value on fields.

**Comment by xiaoxmeng** [comment]
> @xiaoxmeng This PR is very large, > 1K lines. Would it be possible to break it up into smaller pieces? Can the 3 fixed for pre-existing issues listed in "This PR also fixes the following issues:" be extracted into separate PRs? Smaller PRs are easier and quicker to review.

@mbasmanova thanks a lot for the quick review and helpful suggestions. Yeah, I will split them out to help review.

File: velox/exec/HashProbe.cpp [hunk 23]
```diff
@@ -98,33 +124,43 @@ HashProbe::HashProbe(
           "HashProbe"),
       outputBatchSize_{driverCtx->queryConfig().preferredOutputBatchSize()},
       joinType_{joinNode->joinType()},
+      joinBridge_(operatorCtx_->task()->getHashJoinBridgeLocked(
+          operatorCtx_->driverCtx()->splitGroupId,
+          planNodeId())),
+      spillConfig_(makeOperatorSpillConfig(
+          *operatorCtx_->task()->queryCtx(),
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:131]: I wonder if we could remove this parameter from makeOperatorSpillConfig? Since we also pass operatorCtx, the queryCtx can be extract from it. This would make this code a bit shorter and easier to read.

Also, wondering if it makes sense to add operator ID to operatorCtx and make makeOperatorSpillConfig a method of OperatorCtx:

```
spillConfig_{operatorCtx_->makeSpillConfig(core::QueryConfig::kJoinSpillEnabled)}
```

File: velox/exec/HashProbe.cpp [hunk 24]
```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
...
-    if (table_->numDistinct() == 0) {
-      // Build side is empty. Inner, right and semi joins return nothing in this
-      // case, hence, we can terminate the pipeline early.
-      if (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
-          isRightJoin(joinType_) || isRightSemiJoin(joinType_)) {
-        finished_ = true;
+    // 'antiJoinHasNullKeys' flag shall only be set on the first built 'table_'
+    // and we don't expect any spilled probe data at that point.
+    VELOX_CHECK(spillPartitionSet_.empty());
+    spillPartitionSet_.clear();
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:295]: This statement seems redundant as we just checked that spillPartitionSet_ is empty.

File: velox/exec/HashProbe.cpp [hunk 25]
```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
...
+    noMoreInput();
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:296]: I wonder how would we end up in this code path... Won't we finish early upon seeing that hash table has a null key?

File: velox/exec/HashProbe.cpp [hunk 26]
```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
...
+  } else if (
+      (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
+       isRightSemiJoin(joinType_)) &&
+      table_->hashMode() != BaseHashTable::HashMode::kHash && !isSpillInput() &&
+      !hasMoreSpillData()) {
+    // Find out whether there are any upstream operators that can accept
+    // dynamic filters on all or a subset of the join keys. Create dynamic
+    // filters to push down.
+    //
+    // NOTE: this optimization is not applied in the following cases: (1) if the
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:327]: Curious if we have test cases to ensure dynamic filtering is disabled in case of spilling.

File: velox/exec/HashProbe.cpp [hunk 27]
```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
...
     }
   }
+}
+
+bool HashProbe::isSpillInput() const {
+  return spillInputReader_ != nullptr;
+}
 
-  return BlockingReason::kNotBlocked;
+void HashProbe::prepareForSpillRestore() {
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:348]: Finished reading here.

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Read up to HashProbe::prepareForSpillRestore. Flushing commits. Need to take a break and continue reading later.

File: velox/exec/HashBuild.cpp (hunk 9 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp:712]: Yes.

File: velox/exec/HashBuild.cpp (hunk 8 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp:184]: It will run out of memory if the restored partition can't fit in-memory. Comment added.

File: velox/exec/HashBuild.cpp (hunk 10 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp:825]: Added. Thanks!

File: velox/exec/HashBuild.cpp (hunk 8 above)
Comment (mbasmanova) [reply velox/exec/HashBuild.cpp:184]: Would it be confusing to the user having spilling enabled yet still running out of memory. Would it make sense to improve the error message to explain why spilling couldn't help?

File: velox/exec/HashBuild.cpp (hunk 9 above)
Comment (mbasmanova) [reply velox/exec/HashBuild.cpp:712]: Got it. It would be nice to extract into a separate PR.

File: velox/exec/HashProbe.cpp (hunk 23 above)
Comment (xiaoxmeng) [reply velox/exec/HashProbe.cpp:131]: https://github.com/facebookincubator/velox/pull/2917 does this refactoring.

File: velox/exec/HashProbe.h [hunk 28]
```diff
@@ -108,11 +159,83 @@ class HashProbe : public Operator {
...
+  // Used for synchronization with the hash probe operators of the same pipeline
+  // to handle the last probe processing for certain types of join and notify
+  // the hash build operators to build the next hash table from spilled data if
+  // disk spilling has been triggered.
+  ContinueFuture future_{ContinueFuture::makeEmpty()};
+
+  // Used with 'future_' for the same purpose. 'promises_' will be fulfilled
+  // after finishes the last probe processing and hash build operator
+  // notification.
+  std::vector<ContinuePromise> promises_;
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/HashProbe.h:229]: Yeah, updated.

File: velox/exec/HashProbe.cpp (hunk 26 above)
Comment (xiaoxmeng) [reply velox/exec/HashProbe.cpp:327]: Yes, we check dynamic filtering is not set in existing HashJoinTestCase with spilling enabled.

File: velox/exec/HashProbe.cpp (hunk 25 above)
Comment (xiaoxmeng) [reply velox/exec/HashProbe.cpp:296]: getOutput() will finish accordingly and switch to finish state there.

File: velox/exec/HashProbe.h (hunk 18 above)
Comment (mbasmanova) [reply velox/exec/HashProbe.h:105]: Got it. Thank you for explaining. Curious if we have test coverage for recursive spilling.

File: velox/exec/HashProbe.h [hunk 29]
```diff
@@ -108,11 +159,83 @@ class HashProbe : public Operator {
...
+  // Indicates if the operator has more probe inputs from either the upstream
+  // operator or the spill input reader.
+  bool hasMoreInput() const;
+
+  // Indicates if the join type requires to produce the output results from the
+  // join results after all the probe operators have finished processing the
+  // probe inputs.
+  bool needLastProbe() const;
+
+  // Indicates if the join type can skip processing probe inputs with empty
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/exec/HashProbe.h:171]: I see, but do we have code on the probe side to skip writing out these partitions, e.g. avoid storing them into spill files altogether?

File: velox/exec/HashProbe.cpp [hunk 30]
```diff
@@ -189,58 +222,278 @@ void HashProbe::initializeFilter(
...
+  }
+
+  table_ = std::move(hashBuildResult->table);
+  VELOX_CHECK_NOT_NULL(table_);
+
+  maybeSetupSpillInput(
+      hashBuildResult->restoredPartitionId, hashBuildResult->spillPartitionIds);
+
+  if (table_->numDistinct() == 0) {
+    // If the build table is empty, then inner, right and semi joins return
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:306]: nit: skipProbeOnEmptyBuild is a nice descriptive name that doesn't seem to require a comment; would it make sense to move this comment to the header file to shorten this function?

File: velox/exec/HashProbe.cpp [hunk 31]
```diff
@@ -189,58 +222,278 @@ void HashProbe::initializeFilter(
...
+  return spillInputReader_ != nullptr;
+}
+
+void HashProbe::prepareForSpillRestore() {
+  checkRunning();
+  VELOX_CHECK(spillEnabled());
+  VELOX_CHECK(hasMoreSpillData());
+
+  // Wait for table build after prepared for spill restore.
+  auto guard = folly::makeGuard([&] { waitForHashTable(); });
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:353]: nit: waitForHashTable name suggests that waiting for the hash table happens synchronously; I wonder if there is a better name since waiting happens asynchronously. 

nit: the use of guard is a bit surprising here; would it make sense to move waitForHashTable() to the caller?

File: velox/exec/HashProbe.h [hunk 32]
```diff
@@ -108,11 +159,96 @@ class HashProbe : public Operator {
...
+  // remaining to restore. The function will reset the internal states which
+  // are relevant to the last finished probe run. The last finished probe
+  // operator will also notify the hash build operators to build the next hash
+  // table from spilled data.
+  void prepareForSpillRestore();
+
+  // Invoked to read next batch of spilled probe inputs from disk to process.
+  void addSpillInput();
+
+  // Invoked to spill rows in 'input' to disk directly if the corresponding
```
Comment (mbasmanova) [inline velox/exec/HashProbe.h:199]: Perhaps, clarify that this method keeps 'input' as is if no row needs spilling; resets it to null if all rows need spilling; wraps in a dictionary using rows number that do need spilling otherwise.

It is a bit unusual to wrap the top-level row vector in a dictionary.

File: velox/exec/HashProbe.cpp [hunk 33]
```diff
@@ -261,13 +514,31 @@ void HashProbe::addInput(RowVectorPtr input) {
     return;
   }
 
+  if (needSpillInput()) {
+    spillInput(input_);
+  }
+
+  // Check if all the probe input rows have been spilled.
```
Comment (mbasmanova) [inline velox/exec/HashProbe.cpp:521]: nit: perhaps, move this inside of the 'if (needSpillInput())' block.

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Thank you for splitting the PR. It makes it so much easier to review. The code looks great % a few nits.

File: velox/exec/HashProbe.h (hunk 29 above)
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/HashProbe.h:171]: You mean skip spilling the corresponding probe rows? The probe operator will only spill rows if the corresponding partition is set in hash join bridge. Let me just add a test case to ensure that behvavior.

[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/exec/HashProbe.h:171]: I was thinking that if a particular partition happens to be empty on the build side, probe side can just drop rows from that partition without doing anything (e.g. neither spill nor probe the hash table).

File: velox/exec/HashProbe.cpp (hunk 33 above)
Comment (xiaoxmeng) [reply velox/exec/HashProbe.cpp:521]: Good point!
